### ISSUE TITLE
(when ready to release) Update Android app to support IPv6

### ISSUE THREAD

**Issue by danderson** [tailscale/tailscale#1158]
Check that the VPN manager bits of Android aren't angered by the new v6 logic.

**Comment by danderson** [comment]
With https://github.com/tailscale/tailscale-android/commit/d3b57b20cf08e3edf844109447697bc99865a82e applied to update the tailscale.com module version, IPv6 and default routing appear to work correctly on Android. Tested:
 - Log into tailscale on a network with the experimental v6 active.
   - Android app shows v6 addresses for all nodes
   - In `adb shell`, `ip addr` and `ip route show table all` shows v6 addr+routes being correctly added to the network stack
   - Can `ping -6` the phone from another node in my network
 - In admin panel, enable 0.0.0.0/0 and ::/0 routes through another node.
   - Chrome https://ifconfig.co shows IP address changing to that of the egress node
   - In `adb shell`, `ip addr` and `ip route show table all` shows default routes being added for v4 and v6 in the Android VPN routing table.
   - In `adb shell`, `ping google.com` and `ping6 google.com` both work. This is notable because the phone itself (emulator) doesn't have native v6 connectivity on its emulated network, but can still route to the v6 internet via Tailscale.
 - In admin panel, disable default routes.
   - Chrome https://ifconfig.co shows IP address revert back to my local IP
   - The various `adb shell` things confirm that all default routing config has been removed, and I'm back to "local" routing.
 - Reenable default routes in admin panel, toggle tailscale on and off
   - Active default route correctly follows the overall state of Tailscale on the phone.

Leaving this open because there was a minor update required to get tailscale to compile again (see linked commit above) and to enable the v6 routing, which we'll want to do at the latest when we do the Tailscale release that includes v6+default routes. The GUI probably needs to hide v6 addrs as well, which I didn't look into.

cc @eliasnaur FYI, some future android things that will need to be done.

**Comment by eliasnaur** [comment]
> With [tailscale/tailscale-android@d3b57b2](https://github.com/tailscale/tailscale-android/commit/d3b57b20cf08e3edf844109447697bc99865a82e) applied to update the tailscale.com module version, IPv6 and default routing appear to work correctly on Android. Tested:

Thank you.

> The GUI probably needs to hide v6 addrs as well, which I didn't look into.
> 

Can you elaborate? Which v6 addrs, and why do they need hiding?

**Comment by eliasnaur** [comment]
@danderson, thanks for the detailed reply. I've pushed your commit, and a commit that filters IPv6 from the UI. What version of the tailscale.com module shall I release the app with? You selected a version that Go names `tailscale.com v1.1.1-0.20210118213216-9936cffc1a07` in go.mod, but that version is unstable and not allowed by the versioning scripts (`Unstable release v1.1.1-0-g9936cffc1a07 has a non-zero patch number, which is not allowed`).

**Comment by bradfitz** [comment]
@eliasnaur, to get the version number, use `version.sh`:

```
bradfitz@tsdev:~/src/tailscale.com$ ./version/version.sh
VERSION_SHORT="1.3.293"
VERSION_LONG="1.3.293-tdaf2c70a0"
VERSION_GIT_HASH="daf2c70a083aa02496fdc66ec7f7b71f82f9844a"
VERSION_EXTRA_HASH=""
VERSION_XCODE="101.3.293"
VERSION_WINRES="1,3,293,0"
```

Then you should append your git revision to the end of VERSION_LONG, like `1.3.293-tdaf2c70a0-gXXXXXXXXX` where you fill in `XXXXXXXXX` with the git hash of the tailscale-android repo.

**Comment by eliasnaur** [comment]
@bradfitz, it's true that tailscale.com's version.sh returns a different result than tailscale-android's build. However, to determine the tailscale version without git cloning tailscale.com at every build, I'm using the go.mod version.

See [the tailscale-version.sh script](https://github.com/tailscale/tailscale-android/blob/main/version/tailscale-version.sh), [original pull request](https://github.com/tailscale/tailscale-android/blob/main/version/tailscale-version.sh), [previous discussion](https://github.com/tailscale/tailscale/issues/883#issuecomment-723427083).

Did you stop tagging releases, resulting in inaccurate go.mod version strings for tailscale.com?

Potential solutions:

- `git clone` tailscale.com during tailscale-android builds, to get access to version.sh and a repository to query. Slows down edit-build-run cycle considerably.
- As above, but use a dummy version for debug builds. Fast, but then debug builds are subtly different from release builds.
- Tag tailscale.com releases so go.mod version strings match your version scheme, or at least enough to make my `tailscale-android.sh` script work.
- Convert version.sh to Go, so that `go list -m -f '{{.Dir}}' tailscale.com` will include it.

**Comment by bradfitz** [comment]
> Slows down edit-build-run cycle considerably.

So cache the tailscale.com repo somewhere so you don't need to re-clone it per build?

> As above, but use a dummy version for debug builds. Fast, but then debug builds are subtly different from release builds.

That's tolerable. We at least need the `version.sh`-accurate versions for builds that users are running from the Play Store or F-Droid.

But you should still cache the tailscale.com repo on disk somewhere.

**Comment by eliasnaur** [comment]
Thanks, I switched to git cloning the tailscale.com module as you suggested. It still feels weird to bypass the Go module versioning.

I've released "1.3.267-t9936cffc1-g96e2661764c" to Play Store.
